libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1049,31 +1049,85 @@ internal class MapRouteLine(
+                    """.trimIndent()
+                )
+                return
+            }
+            val upcomingPoint = upcomingIndex.point
+
+            if (index > 0) {
+                val distanceToLine = findDistanceToNearestPointOnCurrentLine(
+                    point,
+                    granularDistances,
+                    index
+                )
+                if (distanceToLine > ROUTE_LINE_UPDATE_MAX_DISTANCE_THRESHOLD_IN_METERS) {
+                    return
+                }
+            }
 
             /**
              * Take the remaining distance from the upcoming point on the route and extends it
              * by the exact position of the puck.
              */
             val remainingDistance =
-                traveledIndex.distanceRemaining + calculateDistance(upcomingPoint, point)
+                upcomingIndex.distanceRemaining + calculateDistance(upcomingPoint, point)
 
             /**
              * Calculate the percentage of the route traveled and update the expression.
              */
-            if (granularDistances.distance >= remainingDistance) {
-                val offset = (1.0 - remainingDistance / granularDistances.distance)
-                if (offset >= 0) {
-                    val expression = getExpressionAtOffset(offset)
-                    hideCasingLineAtOffset(offset)
-                    hideRouteLineAtOffset(offset)
-                    decorateRouteLine(expression)
-                }
+            val offset = if (granularDistances.distance >= remainingDistance) {
+                (1.0 - remainingDistance / granularDistances.distance)
+            } else {
+                0.0
+            }
+
+            if (vanishingPointState == VanishingPointState.ONLY_INCREASE_PROGRESS &&
+                vanishPointOffset > offset
+            ) {
+                return
             }
+            val expression = getExpressionAtOffset(offset)
+            hideCasingLineAtOffset(offset)
+            hideRouteLineAtOffset(offset)
+            decorateRouteLine(expression)
         }
     }
 
+    /**
+     * Creates a line from the upcoming geometry point and the previous 10 points
+     * and tries to find the the distance from current point to that line.
+     *
+     * We need to take more points than <previous - upcoming> because the route progress update
+     * can jump by more than 1 geometry point.
+     * If this happens, the puck will animate through multiple geometry points,
+     * so we need to make a line with a buffer.
+     */
+    private fun findDistanceToNearestPointOnCurrentLine(
+        point: Point,
+        granularDistances: RouteLineGranularDistances,
+        upcomingIndex: Int
+    ): Double {
+        return TurfMisc.nearestPointOnLine(
+            point,
+            granularDistances.distancesArray.run {
+                val points = mutableListOf<Point>()
+                for (i in max(upcomingIndex - 10, 0)..upcomingIndex) {
```

Comment:
These points seem to often be the same. It would be nice if we can find a way to take advantage of this to improve performance (in the future).

Reply:
Yea, I wasn't able to find an easy way to do optimization here. The expensive call remains turf itself.